modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/KeysTrackingConfigurationVisitor.java

```diff
@@ -183,5 +195,32 @@ private void endVisit(int previousKeyLength) {
         currentKey.setLength(previousKeyLength);
 
         currentPath.remove(currentPath.size() - 1);
+        legacyNames.remove(legacyNames.size() - 1);
+    }
+
+    /** Calls consumer for all legacy paths. */
+    protected void processLegacyPaths(ArrayList<String> path, BiConsumer<String, String> legacyKeyNewKeyConsumer) {
+        if (path.size() == currentPath().size() - 1) {
+            for (String legacyName : legacyNames.get(currentPath().size() - 1)) {
+                String legacyKey = join(appendKey(path, legacyName));
+                String newKey = currentKey();
+
+                legacyKeyNewKeyConsumer.accept(legacyKey, newKey);
+            }
+
+            path.remove(path.size() - 1);
+
+            return;
+        }
+
+        for (String legacyName : legacyNames.get(path.size())) {
+            path.add(legacyName);
+
+            processLegacyPaths(path, legacyKeyNewKeyConsumer);
+        }
+
+        path.add(currentPath().get(path.size()));
+
+        processLegacyPaths(path, legacyKeyNewKeyConsumer);
```

Comment:
I'm not convinced that this is a right choice, time complexity of visiting the configuration will be too high. Why does this code exist, is it even tested?

The way I see it - there should be two methods, one that does it, and one that does not. And the latter should be private. This code requires some very careful consideration